text/1006-groupped-cli.rst

```diff
@@ -0,0 +1,240 @@
+all seems very limiting. Therefore we propose the following:
+
+* Inspect commands will be exposed **to the CLI** via the ``edgedb inspect``
+  command group.
+
+* Inspect subcommands will be exposed **in REPL** directly, without
+  requiring users to type ``inspect``. E.g. instead of typing
+  ``\inspect list-object-types`` users should be able to continue
+  using the ``\list-object-types`` command.
+
+Overall, the consistency of the CLI is more important than the minor
+inconsistencies of the introspection commands in REPL.
+
+
+Occasional Duplication of Commands
+----------------------------------
+
+Some of the commands will have aliases:
+
+* ``edgedb migrate`` is an alias for ``edgedb migration apply``. The reason
+  for having the alias: this will be a very popular and frequently typed
+  command.
+
+* ``edgedb inspect list-X`` might become aliases for ``edgedb X list``
+  for some types of entities in the future.  While this does not seem like
+  a "pure" solution, there is no harm in having aliases like this.
+
+In general, we believe that having aliases for some commands cannot
+harm the overall developer experience of using the CLI.
+
+
+RFC 1003 -- Rejected Ideas
+--------------------------
+
+The superseded RFC 1003 explicitly rejected the ``<group> <command>`` naming
+scheme, quote::
+
+    * inability to adjust every command naturally in this way;
+
+    * disruptive nature of the change;
+
+    * less verbose ``help`` output; and
+
+    * less natural-sounding commands.
+
+The verbosity of the updated ``edgedb --help`` output can be and will be
+tweaked until it hits the perfect balance of being readable and informative.
+
+The less natural-sounding commands argument is valid, as
+``edgedb create-migration`` certainly sounds more natural than
+``edgedb migration create``. But given that we will likely have between more
+than 30 subcommands, it is clear that giving users a way to organize
+subcommands mentally in categories to memorize the overall structure is more
+important than "making commands sound like plain English".
+
+The proposed change is indeed very disruptive but we believe it is still worth
+implementing it before 1.0. It is important to understand that RFC 1003 was
+written when the CLI had only one "tool within a tool" — ``edgedb server``.
+Since then we have added ``edgedb project`` and it became apparent that we
+will likely continue to add more tools like that.
+
+
+Backwards Compatibility
+=======================
+
+We will supporting all existing commands (e.g. ``edgedb create-migration``)
+until the 1.0 release.
+
+The old commands will be hidden from the ``--help`` output. When run, old
+commands will render a deprecation warning, e.g.::
+
+    $ edgedb create-migration
+    The `create-migration` command has been deprecated.
+    Use `edgedb migration create` instead.
+
+
+Open Questions
+==============
+
+* Do we need to remove or redesign ``edgedb configure``?
```

Comment:
Will rename this to `edgedb config`.

Reply:
While we're on the subject of tweaking the CLI `configure` group, maybe we should also rename `configure insert` to `config add`.

The reasons for the change:
* `insert` sounds like a DB command that often goes together with `delete` and `update`. At least in my experience it did make me look for those other commands in vain and being unsure whether the docs are outdated, or whether the implementation is incomplete. We can still leave the commands in EdgeQL as they are, but at least for the CLI, we might as well avoid the confusion.
* We already are breaking from an exact match between the CLI config commands and EdgeQL `CONFIGURE` commands and we intend to do more of it, specializing `auth` commands into their own CLI group, etc.
* I would argue that we should look at the situation from a perspective of a new user, who isn't aware of all the intricacies of EdgeQL yet. Thus they probably aren't even aware of the `CONFIGURE` commands, they are just trying to figure out how they can change some configuration aspect and `edgedb config --help` is possibly their first stop on this path. Seeing `insert`, `set` and `reset` may be a bit more confusing than `add`, `set`, `reset` when the intent is to either "add" some new config or "update" an existing one. "Updating" or "deleting" configuration is especially interesting in that in the presence of an `insert` command, `set` and `reset` look less like viable candidates for those actions, whereas if the command is a non-keyword `add`, then there's no built-in expectation to match it with other keywords, so `set` and `reset` seem more natural.

I realize that this is very subjective, but I think it's worth considering since even members of our own team got a little confused by the `configure insert`.